src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java

```diff
@@ -89,28 +91,86 @@ public void startStreaming(
+                boolean sendNow = false;
+                for (VEvent event : response.getEventsList()) {
+                    LOGGER.debug("VEvent: {}", event);
+                    switch (event.getType()) {
+                        case ROW:
+                            numOfRowEvents++;
+                            break;
+                        case VGTID:
+                            // We always use the latest VGTID if any.
+                            newVgtid = Vgtid.of(event.getVgtid());
+                            break;
+                        case BEGIN:
+                            // We should only see BEGIN before seeing COMMIT.
+                            if (commitEventSeen) {
+                                LOGGER.error("Receive BEGIN event after receiving COMMIT event");
+                            }
+                            if (beginEventSeen) {
+                                LOGGER.error("Received duplicate BEGIN events");
+                            }
+                            beginEventSeen = true;
+                            break;
+                        case COMMIT:
+                            // We should only see COMMIT after seeing BEGIN.
+                            if (!beginEventSeen) {
+                                LOGGER.error("Receive COMMIT event before receiving BEGIN event");
+                            }
+                            if (commitEventSeen) {
+                                LOGGER.error("Received duplicate COMMIT events");
+                            }
+                            commitEventSeen = true;
+                            break;
+                        case DDL:
+                        case OTHER:
+                            // If receiving DDL and OTHER, process them immediately to rotate vgtid in offset.
+                            // For example, the response can be:
+                            // [VGTID, DDL]. This is an DDL event.
+                            // [VGTID, OTHER]. This is the first response if "current" is used as starting gtid.
+                            sendNow = true;
+                            break;
+                    }
+                    bufferedEvents.add(event);
+                }
+
+                numResponses++;
+
+                // We only proceed when we receive a complete transaction after seeing both BEGIN and COMMIT events,
+                // OR if sendNow flag is true (meaning we should process buffered events immediately).
+                if ((!beginEventSeen || !commitEventSeen) && !sendNow) {
+                    return;
                 }
 
-                Vgtid newVgtid = getVgtid(response);
-                int numOfRowEvents = getNumOfRowEvents(response);
+                if (numResponses > 1) {
+                    LOGGER.info("Processing multi-response transaction: number of responses is {}", numResponses);
+                }
+                if (newVgtid == null) {
```

Comment:
Yes, sometimes there is no vgtid and it's expected, for example in the case of [BEGIN, COMMIT] as we see. In this case, we want to simply skip such events, because they cannot do anything (neither rotate vgtid or sending new recorsd as a transaction). 

BEGIN/END messages with null transaction id cannot be sent. However, with current untouched code, if we process them, it will make [BEGIN, COMMIT] to use the vgtid of the previous transaction, which is what we want to avoid. Therefore I'd better just skip it here, wdyt?